core/services/vrf/v2/listener_v2_log_listener.go

```diff
@@ -0,0 +1,451 @@
+	"github.com/smartcontractkit/chainlink/v2/core/services/vrf/vrfcommon"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/mathutil"
+)
+
+func (lsn *listenerV2) runLogListener(
+	pollPeriod time.Duration,
+	minConfs uint32,
+) {
+	lsn.l.Infow("Listening for run requests via log poller",
+		"minConfs", minConfs)
+	ticker := time.NewTicker(utils.WithJitter(pollPeriod / 2))
+	defer ticker.Stop()
+	var (
+		lastProcessedBlock int64
+		startingUp         = true
+	)
+	ctx, cancel := lsn.chStop.NewCtx()
+	defer cancel()
+	for {
+		select {
+		case <-lsn.chStop:
+			return
+		case <-ticker.C:
+			start := time.Now()
+			lsn.l.Debugw("log listener loop")
+			// Filter registration is idempotent, so we can just call it every time
+			// and retry on errors using the ticker.
+			err := lsn.chain.LogPoller().RegisterFilter(logpoller.Filter{
+				Name: fmt.Sprintf("vrf_%s_keyhash_%s_job_%d", lsn.coordinator.Version(), lsn.job.VRFSpec.PublicKey.MustHash().String(), lsn.job.ID),
+				EventSigs: evmtypes.HashArray{
+					lsn.coordinator.RandomWordsFulfilledTopic(),
+					lsn.coordinator.RandomWordsRequestedTopic(),
+				},
+				Addresses: evmtypes.AddressArray{
+					lsn.coordinator.Address(),
+				},
+			})
+			if err != nil {
+				lsn.l.Errorw("error registering filter in log poller, retrying",
+					"err", err,
+					"elapsed", time.Since(start))
+				continue
+			}
+
+			// on startup we want to initialize the last processed block
+			if startingUp {
+				lsn.l.Debugw("initializing last processed block on startup")
+				lastProcessedBlock, err = lsn.initializeLastProcessedBlock(ctx)
+				if err != nil {
+					lsn.l.Errorw("error initializing last processed block, retrying",
+						"err", err,
+						"elapsed", time.Since(start))
+					continue
+				}
+				startingUp = false
+				lsn.l.Debugw("initialized last processed block", "lastProcessedBlock", lastProcessedBlock)
+			}
+
+			pending, err := lsn.pollLogs(ctx, minConfs, lastProcessedBlock)
+			if err != nil {
+				lsn.l.Errorw("error polling vrf logs, retrying",
+					"err", err,
+					"elapsed", time.Since(start))
+				continue
+			}
+
+			// process pending requests and insert any fulfillments into the inflight cache
+			lsn.processPendingVRFRequests(ctx, pending)
+
+			lastProcessedBlock, err = lsn.updateLastProcessedBlock(ctx, lastProcessedBlock)
+			if err != nil {
+				lsn.l.Errorw("error updating last processed block, continuing anyway", "err", err)
+			} else {
+				lsn.l.Debugw("updated last processed block", "lastProcessedBlock", lastProcessedBlock)
+			}
+			lsn.l.Debugw("log listener loop done", "elapsed", time.Since(start))
+		}
+	}
+}
+
+// initializeLastProcessedBlock returns the earliest block number that we need to
+// process requests for. This is the block number of the earliest unfulfilled request
+// or the latest finalized block, if there are no unfulfilled requests.
+// TODO: add tests
```

Comment:
is this TODO still undone?

Reply:
Being done as part of this PR: #11299 but may not be merged until afterwards